### PR TITLE
feat: use mongodb official driver instead of mgo

### PR DESCRIPTION
**Pull request by kevwan**
Tasks:
- [x] add `mon` package for `mongodb` operations
- [x] add `monc` package for `mongodb` operations with cache enabled

Notes:
* keep `mongo` and `mongoc` for couple releases
* change `goctl` to generate `pg` models with new package after couple release, depends on feedbacks

We'll probably provide migrating command to do the migration automatically.

### CODE REVIEW

File: core/stores/mon/collection.go [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+	mopt "go.mongodb.org/mongo-driver/mongo/options"
+)
+
+const defaultSlowThreshold = time.Millisecond * 500
+
+// ErrNotFound is an alias of mongo.ErrNoDocuments
+var ErrNotFound = mongo.ErrNoDocuments
+
+type (
+	Collection interface {
```
Comment (fynxiu) [inline core/stores/mon/collection.go:21]: Collection should have comment

File: core/stores/mon/collection_test.go [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+		c := decoratedCollection{
+			Collection: mt.Coll,
+			brk:        breaker.NewBreaker(),
+		}
+		mt.AddMockResponses(mtest.CreateCursorResponse(
+			1,
+			"DBName.CollectionName",
+			mtest.FirstBatch,
+			bson.D{
+				{Key: "_id", Value: primitive.NewObjectID()},
```
Comment (fynxiu) [inline core/stores/mon/collection_test.go:137]: Is this line of code redundant?

**Review by fynxiu — APPROVED** [review]
lgtm

**Review by chenquan — CHANGES_REQUESTED** [review]
`core/stores/mon/utils_test.go` -> `core/stores/mon/util_test.go`

File: core/stores/mon/collection.go [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+	}
+}
+
+func (c *decoratedCollection) Aggregate(ctx context.Context, pipeline interface{},
+	opts ...*mopt.AggregateOptions) (cur *mongo.Cursor, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		starTime := timex.Now()
+		defer func() {
+			c.logDurationSimple("Aggregate", starTime, err)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:119]:
```suggestion
		defer c.logDurationSimple("Aggregate", starTime, err)
```

File: core/stores/mon/collection.go [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) BulkWrite(ctx context.Context, models []mongo.WriteModel,
+	opts ...*mopt.BulkWriteOptions) (res *mongo.BulkWriteResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDurationSimple("BulkWrite", startTime, err)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:133]:
```suggestion
		defer c.logDurationSimple("BulkWrite", startTime, err)
```

File: core/stores/mon/collection.go [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) CountDocuments(ctx context.Context, filter interface{},
+	opts ...*mopt.CountOptions) (count int64, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDurationSimple("CountDocuments", startTime, err)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:147]:
```suggestion
		defer c.logDurationSimple("CountDocuments", startTime, err)
```

File: core/stores/mon/collection.go [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) DeleteMany(ctx context.Context, filter interface{},
+	opts ...*mopt.DeleteOptions) (res *mongo.DeleteResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDurationSimple("DeleteMany", startTime, err)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:161]:
```suggestion
		defer c.logDurationSimple("DeleteMany", startTime, err)
```

File: core/stores/mon/collection.go [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) DeleteOne(ctx context.Context, filter interface{},
+	opts ...*mopt.DeleteOptions) (res *mongo.DeleteResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDuration("DeleteOne", startTime, err, filter)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:175]:
```suggestion
		defer c.logDuration("DeleteOne", startTime, err, filter)
```

File: core/stores/mon/collection.go [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) Distinct(ctx context.Context, fieldName string, filter interface{},
+	opts ...*mopt.DistinctOptions) (val []interface{}, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDurationSimple("Distinct", startTime, err)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:189]:
```suggestion
		defer c.logDurationSimple("Distinct", startTime, err)
```

File: core/stores/mon/collection.go [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) EstimatedDocumentCount(ctx context.Context,
+	opts ...*mopt.EstimatedDocumentCountOptions) (val int64, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDurationSimple("EstimatedDocumentCount", startTime, err)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:203]:
```suggestion
		defer c.logDurationSimple("EstimatedDocumentCount", startTime, err)
```

File: core/stores/mon/collection.go [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) Find(ctx context.Context, filter interface{},
+	opts ...*mopt.FindOptions) (cur *mongo.Cursor, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDuration("Find", startTime, err, filter)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:217]:
```suggestion
		defer c.logDuration("Find", startTime, err, filter)
```

File: core/stores/mon/collection.go [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) FindOne(ctx context.Context, filter interface{},
+	opts ...*mopt.FindOneOptions) (res *mongo.SingleResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDuration("FindOne", startTime, err, filter)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:231]:
```suggestion
		defer c.logDuration("FindOne", startTime, err, filter)
```

File: core/stores/mon/collection.go [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) FindOneAndDelete(ctx context.Context, filter interface{},
+	opts ...*mopt.FindOneAndDeleteOptions) (res *mongo.SingleResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDuration("FindOneAndDelete", startTime, err, filter)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:246]:
```suggestion
		defer c.logDuration("FindOneAndDelete", startTime, err, filter)

File: core/stores/mon/collection.go [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+}
+
+func (c *decoratedCollection) FindOneAndReplace(ctx context.Context, filter interface{},
+	replacement interface{}, opts ...*mopt.FindOneAndReplaceOptions) (
+	res *mongo.SingleResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDuration("FindOneAndReplace", startTime, err, filter, replacement)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:262]:
```suggestion
		defer c.logDuration("FindOneAndReplace", startTime, err, filter, replacement)
```

File: core/stores/mon/collection.go [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) FindOneAndUpdate(ctx context.Context, filter interface{}, update interface{},
+	opts ...*mopt.FindOneAndUpdateOptions) (res *mongo.SingleResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDuration("FindOneAndUpdate", startTime, err, filter, update)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:277]:
```suggestion
		defer c.logDuration("FindOneAndUpdate", startTime, err, filter, update)
```

File: core/stores/mon/collection.go [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) InsertMany(ctx context.Context, documents []interface{},
+	opts ...*mopt.InsertManyOptions) (res *mongo.InsertManyResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDurationSimple("InsertMany", startTime, err)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:292]:
```suggestion
		defer c.logDurationSimple("InsertMany", startTime, err)
```

File: core/stores/mon/collection.go [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) InsertOne(ctx context.Context, document interface{},
+	opts ...*mopt.InsertOneOptions) (res *mongo.InsertOneResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDuration("InsertOne", startTime, err, document)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:306]:
```suggestion
		defer c.logDuration("InsertOne", startTime, err, document)
```

File: core/stores/mon/collection.go [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) ReplaceOne(ctx context.Context, filter interface{}, replacement interface{},
+	opts ...*mopt.ReplaceOptions) (res *mongo.UpdateResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDuration("ReplaceOne", startTime, err, filter, replacement)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:320]:
```suggestion
		defer c.logDuration("ReplaceOne", startTime, err, filter, replacement)
```

File: core/stores/mon/collection.go [hunk 18]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) UpdateByID(ctx context.Context, id interface{}, update interface{},
+	opts ...*mopt.UpdateOptions) (res *mongo.UpdateResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDuration("UpdateByID", startTime, err, id, update)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:334]:
```suggestion
		defer c.logDuration("UpdateByID", startTime, err, id, update)
```

File: core/stores/mon/collection.go [hunk 19]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) UpdateMany(ctx context.Context, filter interface{}, update interface{},
+	opts ...*mopt.UpdateOptions) (res *mongo.UpdateResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDurationSimple("UpdateMany", startTime, err)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:348]:
```suggestion
		defer c.logDurationSimple("UpdateMany", startTime, err)
```

File: core/stores/mon/collection.go [hunk 20]
```diff
@@ -0,0 +1,418 @@
...
+	return
+}
+
+func (c *decoratedCollection) UpdateOne(ctx context.Context, filter interface{}, update interface{},
+	opts ...*mopt.UpdateOptions) (res *mongo.UpdateResult, err error) {
+	err = c.brk.DoWithAcceptable(func() error {
+		startTime := timex.Now()
+		defer func() {
+			c.logDuration("UpdateOne", startTime, err, filter, update)
+		}()
```
Comment (chenquan) [inline core/stores/mon/collection.go:362]:
```suggestion
		defer c.logDuration("UpdateOne", startTime, err, filter, update)
```

**Review by chenquan — CHANGES_REQUESTED** [review]

File: core/stores/mon/collection.go (hunk 3 above)
Comment (kevwan) [reply core/stores/mon/collection.go:119]: It's incorrect, the defer call depends on the err value. the `err` might be changed later.

**Review by chenquan — APPROVED** [review]
lgtm

**Comment by kevwan** [comment]
> `core/stores/mon/utils_test.go` -> `core/stores/mon/util_test.go`

done.

**Pull request merged**